Support dynamic line heights using decorations

remcohaszing opened the pull request:
This adds support for dynamic line heights using decorations in Monaco editor. This is done by adding the `lineHeight` property to `IModelDecorationOptions`. This property represents the line height of the decoration all lines it touches. The line height can only be increased, not decreased.

For testing, I used the following JavaScript and CSS in the Monaco playground:

```js
var jsCode = [
	'"use strict";',
	"function Person(age) {",
	"	if (age) {",
	"		this.age = age;",
	"	}",
	"}",
	"Person.prototype.getAge = function () {",
	"	return this.age;",
	"};",
].join("\n");

var editor = monaco.editor.create(document.getElementById("container"), {
	value: jsCode,
	language: "javascript",
});

var decorations = editor.createDecorationsCollection([
	{
		range: new monaco.Range(3, 1, 5, 1),
		options: {
			lineHeight: 30,
			isWholeLine: true,
			linesDecorationsClassName: "myLineDecoration",
		},
	},
	{
		range: new monaco.Range(7, 1, 7, 24),
		options: {
			lineHeight: 30,
			inlineClassName: "myInlineDecoration",
		},
	},
]);
```

```css
.myInlineDecoration {
	color: red !important;
	cursor: pointer;
	text-decoration: underline;
	font-weight: bold;
	font-style: oblique;
	font-size: 30px;
}

.myLineDecoration {
	background: lightblue;
	width: 5px !important;
	margin-left: 3px;
}

```

Since text might now not fill the entire line height, I chose to align the text at the bottom of the line. This matches default behaviour of typical document processors such as LibreOffice Writer and Google Docs.

![image](https://github.com/microsoft/vscode/assets/779047/0effc3af-fc21-452e-a88d-8b8f9564bd25)

The following screenshot shows that the following works:

- Line decorations are full height
- Text is aligned at the bottom
- The cursor is aligned correctly
- The current line indicator is adjusted to the line height
- Matching words highlighting is aligned correctly

![image](https://github.com/microsoft/vscode/assets/779047/1a9e58c8-d509-4aac-805e-da948ce0a0f1)

The following screenshot shows matching brackets are aligned correctly.

![image](https://github.com/microsoft/vscode/assets/779047/45498bf7-352a-4a56-bed8-ba061d952e12)

The following shows that text selection works. It also shows that the height of the selection indicator is adjusted to the editor’s line height, not that of the decorator. This is different from for example how LibreOffice does it. The same applies to other types of highlights, such as matching words and search.

![image](https://github.com/microsoft/vscode/assets/779047/7409e75e-04ef-4e96-b0cf-597d03c4a61b)

Text selection in LibreOffice

![image](https://github.com/microsoft/vscode/assets/779047/8ee10338-6a8f-4da7-bc54-e5290c578d83)

Things that still need to be done:

- [x] Things are still weird if I insert or remove new lines.
- [x] Discuss if highlights should use the editor line height or the line’s line height. I went with using the full line height, as my client requested this, and this is consistent with LibreOffice and Google Docs.
- [x] Fix handling of inserting new lines.
- [x] Fix handling of removing new lines.
- [x] The search box breaks layout.
- [x] Fix layout shifts height when scrolling.
- [x] Handle word wrap
- [ ] Handle cursor size

cc @hediet 

Closes #147067

remcohaszing committed:
Support dynamic line heights using decorations

hediet commented on src/vs/editor/browser/view/dynamicViewOverlay.ts:
```diff
@@ -7,6 +7,7 @@ import { RenderingContext } from 'vs/editor/browser/view/renderingContext';
 import { ViewEventHandler } from 'vs/editor/common/viewEventHandler';
 
 export abstract class DynamicViewOverlay extends ViewEventHandler {
+	public lineHeight?: number | undefined;
```
This looks fishy. Why does this need to be here?

hediet commented:
Nice work so far! I didn't take a very close look, but keep in mind that if no line heights are set, there shouldn't be any performance hit by this feature.

remcohaszing committed:
Remove unrelated changes

remcohaszing committed:
Use new view-overlay CSS class

remcohaszing committed:
Fix updating lines when typing

remcohaszing committed:
Remove dead code

remcohaszing committed:
Cache the result of getDecorationsLineHeightMap

remcohaszing committed:
Fix typos

remcohaszing committed:
Use Uint8ClampedArray to store line heights

This limits the maximum line height to 255, which seems sufficient.

remcohaszing committed:
Fix some line height related calculations

remcohaszing committed:
Fix scrolling issues

remcohaszing committed:
Fix the cursor position

remcohaszing commented:
I created another fun little test case:

```js
const model = monaco.editor.createModel(`# Coming soon

## to VSCode

### (and Monaco editor)

#### Variable line heights

##### Allowing for fancy stuff

###### Such as bigger markdown headers

And this is regular text`, 'markdown');

// Hover on each property to see its docs!
const ed = monaco.editor.create(document.getElementById("container"), {
	model,
	automaticLayout: true,
});

var decorations = ed.createDecorationsCollection([
	{
		range: new monaco.Range(1, 1, 1, Infinity),
		options: { lineHeight: 50, inlineClassName: "m50" },
	},
	{
		range: new monaco.Range(3, 1, 3, Infinity),
		options: { lineHeight: 45, inlineClassName: "m45" },
	},
	{
		range: new monaco.Range(5, 1, 5, Infinity),
		options: { lineHeight: 40, inlineClassName: "m40" },
	},
	{
		range: new monaco.Range(7, 1, 7, Infinity),
		options: { lineHeight: 35, inlineClassName: "m35" },
	},
	{
		range: new monaco.Range(9, 1, 9, Infinity),
		options: { lineHeight: 30, inlineClassName: "m30" },
	},
	{
		range: new monaco.Range(11, 1, 11, Infinity),
		options: { lineHeight: 25, inlineClassName: "m25" },
	},
]);
```

```css
.m50 { font-size: 50px; }
.m45 { font-size: 45px; }
.m40 { font-size: 40px; }
.m35 { font-size: 35px; }
.m30 { font-size: 30px; }
.m25 { font-size: 25px; }
```

remcohaszing commented:
As far as I know everything works now.

Tests are still missing (old ones commented out). I would like to fix that once it’s confirmed the rest looks good.

remcohaszing committed:
Make text wrapping decorations full height

This applies to selected text, search highlight, and same word
highlight, and other forms of highlighting.

remcohaszing committed:
Handle line wrapping

hediet commented on src/vs/editor/browser/viewParts/viewCursors/viewCursor.ts:
```diff
@@ -165,7 +165,7 @@ export class ViewCursor {
 			}
 
-			const top = ctx.getVerticalOffsetForLineNumber(position.lineNumber) - ctx.bigNumbersDelta;
+			const top = ctx.getVerticalOffsetForLineNumber(position.lineNumber + 1) - this._lineHeight - ctx.bigNumbersDelta;
```
What is the reason for this change? Shouldn't it be the same value?

hediet commented:
[This looks very nice](https://microsoft.github.io/monaco-editor/playground.html?source=http%3A%2F%2Flocalhost%3A5001%2Fout%2Fvs#XQAAAAIyHwAAAAAAAABBqQkHQ5NjdMjwa-jY7SIQ9S7DNlzs5W-mwj0fe1ZCDRFc9ws9XQE0SJE1jc2VKxhaLFIw9vEWSxW3yscx5R7OoJuetpg8d5BCIdogK6CBmgyKHIOVjH6uO-oA-RA-rcpQdGBlgwMT7NDeGjyzFj3-tzLvKPER4KU1q5CFU4o8OuWxPugGXRs81sjnzAagm0SZ8wjqvI-s5u8bIbNjlG2C-vcJEEQTNPBVakvl1p4BvWf7ugzf719hQH64EmV-ppumxx0PnQ6jZ60WvLLcNczyPYuRcOQ9dlj5SXLNXSg0y5k7cpq2U5SibYQQW4QweLpNbBru642-O6GUbIaFjkJ8DXu3LW6CbyNjuVcN9x2yyr0kULTqzD4sRuMYtF8b8H6c51IbGifDHA64Tqi71lxe4Z0LxTmil8qeVDpbHODxmbk5XIlWjZfDp2ZChheJHqfIE9A149s3KAqdGg0I2Z2Aduoz0Z7xH7FbnOiGVHL4KnLRC0KlAPYdOAMyA6ZihHpJ78mHxM6zEbFKIdqbYTR1aVXPpPnDlYYCP7w7AeeFbzN9ruPzp8FRASG8twtf2gdyOI6JM2sXFriHJCDUbYczJyu8PkeUCiHP4zHzsxLZziaziPZvdRwHuOk4rGDD_ohJhJJmdXmgiwMJq2udHvZvALWpDJ7sMPLIyCfWtnaTRIn9hwuPk3mklHpQgVPh307A75RddZyXSGMr7qNeiXQtUd-G0ERB-7ZXXVdDizsDvTj7_0BybsvDrzKv5sglozJVvAmS2f-2iVWDgdSJInPu7fOA3wF4kku4NccVTDbws7GE6N6WPixA6kX9QNan7OTx8bavR4HUZYyiDbxvdABivIrmPW1rG4xey_bztTTAaEWwV7x-TTSbKOozBueP9hcT5a5aMdkPGpzaJOJMzo3Cp_q_bIUpKMxz0KiFhQ_qGPaVe2Cepph5L6Fk-cojKbLs0Lqk4ltIkMoSMgolOkcBb_8a1rcp_H3TBj7jARab6hHgrhLTWAz2qhe-To60xGSw3mo2vMRPqaN_vpGY5h6IzZRX308JUEnXPyTjpnBWpBXZ7XigxeEmS96pIPTsMAb8rUiklRf-yHYuiak6DdQGahJfdSUGkRr1f-wVZFgjavPJy3Gkrkv5UoHU9maWQujxkFR0MRZuNHs7LqiN-ajnBQwhewiuiY96tJVRaNlcgg-ir2IAYSxtfLxoPYI4nI344S5i-Ky6Rz-vyM8XPyO6nXKy7wz2Qwm06rM9TbvJRx6sEf0yE_BPUjGV3yTP-s5bZSk-skulk4200a72AHPtY_T74hnbX_m9doxelyngWgg3YFaefPDn2JdgW9MBYF0iNRWaGg9L9yFoB2uSV1RjNMyRHbX9C0VtKui614VXGv--gjI2mN_Z7UlVSifn6vgnt2DVN5gKYuO2CI_W1DEoai0p_Jk2AzK-d_uAAW66fR1D1os0qjHH-Wf4nXxV0n1SPZAZpDRTsdU5XmL59J1D6T0ktGhaeA-3QaGxmKJ6oOFHzxtikDIUzosJpvtxg0cuASvR3S0h83N9wNTN3-pwB6YePz9_DlRAuuleBDdJv31dV1_VBjFhGyMJ92e91YdrlGEqy_tU0pRJJIU5niHexh-fWhj4j9_GU7tLgJbzn7dp24iZtnqDu1E_9E-HQqMeONk5BNHO-0ITY89-owkK9Yw7Tw_EFcEWOXL60W80sBE2ypuR2fLYett5U8XlrAHav1-F5FS45OpEyqWAz-4Kr9AkVYXp_ZYvUEuecTFng7Lde1qV-iXOCxua9g4gQMUSgH45crIJnNrDTu-SILgkAhuhjGmBnhSoK2LZMN7IVIKZHTT1aDP1h7kBTD9bMoUoRF_15q54cz6590UFJeRrmSR_KK8vFHYCLxBkzTlGkWckYkV4pdRMew3FdM4N0YhHchYNLz4H28OeK6pzLvR1KZZV47XkcFR31rbYnQV7B-kWHs5EZwjRbZ3atz76WzJQavwf2ezcyGwClaF-YnB8km-LZpH6fVcqabsS4ixRJbzwkRBklpXZeStxktXoejuUCS46FmQe6bz3i1qDJFLAF69UHvYC3Xifzmt_fejGN9I5weJTYYkMXPlhD4_yp5-0th-zh2BXFvSsfgxQ4fGXeRLpAEzD4dvdWVwd1nygpuYMTNZG8UpmDN7U19N01YXTDW5ar5U1IU5W7u19NDnKHBjO3yg6rplDGT0xwqEzp6zmOhAqdx8-x75MIfyi7muLb3Pr5es8wZD7DxYoTNyGpThI-EkuhLJOzpALRDobR5vSJ_qoHBCxzS_Q8YJ9W09uJmmSKK3dWD_QfxEk2zaFwSArbd4UQT1obaL8lwqxuu6oP2emehxGAChlKYBA8uPQ3kVPrJ45qddbAM0rHTgGpzmE_EaHvxSmo9Nqx-SR7abaaKsLYv71etn1tMLvSOndeOon26BxVrzaQmxfYDzHFmZEKzvOY6KDkqSKbwvyhlREkZJM1zs7qNlbJILC50S72HJrQiLi28WNDouzsJm9ItcLffYN4y4HLVcDH4VB4Hbf0o6XHUs8peLFkORU1Y9G_exJRpRGWqMp7Bomqc8C4ecMz9bKEl51UxnnoKMI-JGeza5uCAlpbg1vn4yzZJtRbgWenhD2ERvI0fnnRz2yRlAmCNYoqtEIYiokYzlZM11LxGH91Qy_kF29nTwXmbdPDBjgRNXuFcP8i6yAATPlYb2GqFEeXZQu5w1gvS_2zJOe9jDj74SED5BYHyu81h06lbrADaRgqTUTUTULvd0sd364rKqakuzJiN4XjUkc5-i_yFmavapEBtPtvFppT5lpwM1ZAPmjckMtviF54uP0To4mJFIW6WrvzUf5gOVsYCm7ZDpoXCzSCr4EwbnV4RPOqoK25IoLkfMMTwiO3oHaEqBB65a-JxfZNjdvm1bmV9nUM4lWTfSA9lXxHs0hb3GTKLrxy96iqSfMuiaHNC3D7HeebpnxvqjM3tOtPLNDbFv-2AjqHE9A8V_VDay80RKm7TKXM3lwICqLsnktWTJ2kmOvWVIYDmZFHMxhx2yQrG12XxL8GhjsvHFVSFNL-74NJOOH7V3E9Kc15hoOIF8b2kBwYtF1qEBYtk9k6OQbt-4_4lOtu8wdjyBvDuqGsba-s1XWqNEZZgZpRm8ych2sMRFkDtBpDy5jkJ4DBbuSOGYFANpOkNVkxdVvK7Z-ljqi2I4nECHiZQucZEyvavIxyeOQbjsbxm5kQ52Ph5s981Y6klyuvuC0RhloPXKlNoAlaiN2jxLPMbGZlOg1qjKA0KC1lhNZjMp58hbDl1GghK_1d_-6pqdifGd4kP5o_p5JBKyPac_y_ifUlxxhL5lNQCoAKLg96YxRSlwlJrYneNCQ2EF3PAv5vXaoA7JQVmnKTQa2BWjMWGndao0XOhWlQv6_tqBH3BGhUhPa05uIZltjDn4LoyGSopnBJbsVkzoPNv3kCL_ThBOQG8L0jBtRL7e8DoxX_YOwBVwIUE4RSm2PsH79iI94O0ZaeqiZc4utK2JqysFTnZIcXqmgZgZ1eA6eZXusRZliWKXILbDWUTZet3OieHcMkMWoZ6JanCkGHpKhkQnm5yjt4LvHOfMhhwPMixku-M4eeYejyOHsg3Ccd4zHcJOTMhY2aM4ufxZURN6n7BUWQ-LOEIyDghagh1nSR-DkzTLQYSPgQPkwZKrOueN984LHU8J3v93mpim3PpnbLsxIhDRqqW79hIQMaENXM3EjEbpfZoVsVGAFqAvmv2MJZfxznAlz9d1pKHnSETr6btry3m-BGm2Qhc7GHjb2j_NkHxhHdguPXbsLpzy1DzPyAN574B7khx88p_o71sjKlV-aJNEFFTb0Kv0nOx00gtgHYds6_6sWme2FWfX3D6W0qbu5BU6c3d1sTi06yBzBo1eCuPFr6Ja_FaGl_WnXLcxV9f1I7qtXNa-vI5TeFm81DuipDKbviUQGgF6zFp0l4rCPyDu8onNTIcLgtYNxXvGhgcLZSFvjT9v9aqWEFEBeiBI2RICHuXnItC-Nk4UgtvukvFk9dgbWp-jiDlQILRVHxSB5BWsZOkx8HrTgJ_6nNgCk0yZrsyNyTDzrfexDKEfXFPsjo-daEfuVox5ZueWpqlzIlv_Fmwc3)

![chrome_WUw9gbCdfO](https://github.com/microsoft/vscode/assets/2931520/30547374-70a5-4a49-9b35-d3fcbe303650)

[Another demo:](https://microsoft.github.io/monaco-editor/playground.html?source=http%3A%2F%2Flocalhost%3A5001%2Fout%2Fvs#XQAAAAILegAAAAAAAABBqQkHQ5NjdMjwa-jY7SIQ9S7DNlzs5W-mwj0fe1ZCDRFc9ws9XQE0SJE1jc2VKxhaLFIw9vEWSxW3ysc1TQCNs_SwgXffaDq9aGbpI7-lB9zM7jgKMR9NQBC7tEv5g68wm21--i4onfeikIAGMKjzdVf-L-BfQ1w56j7l9MFdb09F_DHTAkq_L9FMo6c4AQ3wYYsTZgs2rnH9obY0Y9G3i62YIKwkLhYP000n6SjQokIi08QdxrIYIEqaP55EkS-i4v7CG1RUEHLGpA2Sl9TFtWCMBauYomUYH2KYmDQ69byClN0NJRrNkyFFxync-vVMqrS_pw3D0l5VksSS_PRrooEmph8BHswnjsl4g9CZXq1P-5a9DYWEATM6arxC6iZFdbQgI3KHpaoYTQX07i_0aT1HWF-cZ0M6mDs9F1byxG8zeJzQFRAdsikwbvfRvN4V3Oy045KwCzJ3pUiUJaka3mWCNE-GWeFo4tP72yFEeAEQ6FgR-pOM8E0qGyKP0P3mbi6lc_Q2MczFCdmbunNW5vh4upWGDemveUn-4DwtJyzC5cUbR6IaLRxnCVPm5K69asTNg7oA-BjmMgrdjmSA4hMMGUKyh_L3JhJsHL-PhaRqzhewN-sqlE3k0IkMj6Kv28Ix7UvRFwL_pMAHnWEfSTEQEcZ07Wbq791rlqPfK6wcyC5ZyiMdsRsyFGoJzQUyHrdLGeUUAgroLGW2VGvMWqDsw6eJUER1vkthrT_a4-dPNjSE7OccFQ2I6vAZyBPzCkcBLa9Ly0yT-G243JXhAlNADY0fyIGD1OrNQeb_9mLmAZ-D9Yu-SJV9lK-sfT0FQloZCUhIDUu_odjegtJ5_bVppRZuPLtj5hjGCbgoFFRC1632x-_UgaFym1wcVEtE48x2V6HdXnFkgRohyR2UQ84rbD883aAOHQhnqFTouk67tgcKXOy_7Hvumlp5MGr6ANHV-8r34zJGS8_O0cGcCHSH3-6KqiGrScMQ4vbUjnUfpwHVHXo6aMWJc2ArP-h5igF7GlI4xfQO1DV-Qc4uQ6JEYJ8iGvRIMSKa_BRPGc8YLOJykFmPSPoVQ1pQn8kn43J_JckAY0Qo_UnRdWI7PRwcW8_IHcZalrfJ7cpuRPobz46UWADZMbBscrPfjNu5-LujQoKhqASHZtHcKDD03xEJ3MwoUbBi_Vq3yDzaFFPTkSW2_Lhp-NX7VroW7GfiwP1yDgRSpjtEnnENMQ_iAILKmR4Cvwos8NfCS_GsmTjMARYBAKPZVjvmqT4V9g0OK1TKqXm-DIudrs8F8iSqKy-gno8R4WHE-ML08WBZU_8-hEhUtTxkiWTwpyyHs2oA3CWTOr9Ws5QhQFmrJBKF0gO6qz2Edn-smXfrbAzWn_np15AE5WlMCTs0MqxVust3G7Yi3JC-JNDrMmPxWSKPYgUd-gLNkL1HgVzAvYBPB-LP-6U6OCFry57hEyGb_HwOCTlGKAn8bEx-KTf_U5fiBeEOYvXNUaqOqVXN1IHz-cuq7T0a7UZjTQZZ8m-7ndfigTV94aTZzKBZMxCAKyFkl1rhZctF2lfDP0bqohntKNOb8wXKDDUWfHUEMCaka9b--pZv2mej--gXgL1daCMWrbIFcGzxXOXaErD5Mql4796AJfpQsdFfR3FkiVvZ1uWRtzwDmg18lyzwfnCW4mlhAVo2f4nxW3RE4CAAYieJhw2te3g2y8cJaqLZIDXrHxbcL4bEdhmp73Q_pxpANf78NrX2GLG1U1uLVivQy-soSinUgp9IcXRuDssUbrnqn09jwzUxsEJ-r8zAhQ7eJckQVWUpt0HPZLpUeWEjRa9QAT3H1rHfW2R0vuma2Iy0Y3kqQFSiE9SJmG9QcxT87UMjEONcGcGHLlPNEz95dt_XfofL_GQfq2862xSLEH7EHFLzTUQsDNWTlj_AztHfvUn4It3nTwNxB8uSTnmkHvBXVSwfmuQarx2nKfQJQutC_d8jZoFS33R_btaXs4ndhXk5CzinDZ84W8qJe0Jkz1z6C0Sm5Q6Scqy1GvczQc3PIk0Dbgyfm-5aFpHsJGN8BIIuimn9H2Um0Jk34rPlIjABnAHzEqNQLSvDUd8nGBni25I7rzWJiSYH_iAqkesOysxQAHVwpRTAfqhFSxpaEHIQK74Rnq1-smPo-MaSYpA6cCOeMsZJXtaFu4DxcC1R84TSkNV2W0uHufvSvJbcPukAfTXIaqMyg_LgZAsn26N2_AL6Bqt1736hpAG_XmrYP9j_kemw58cwMizfe2ShJePzO_wQfczd-qEAJNpcc06FBwSrHzxy2DcbxzPkxheMBYTgw2nGmFJ-s4C-y46hpiriQHUv4jZ4qS2L6FBfmZ33nYrmA3kkD1MKiLEo6Zsp5Dksn6MQ6x3P_jXs5LkeMZwnjBIACiM3y3g1TpkC8JBnBU9VFZ_vf0sm7Wyr2ZlowG19QAodsdhpGN29cz4opj4fjjNiK2bj4wYJFZpBzWkvdmxXUEMDU49rY64zqvQPOMDmcNwVI-9poE9aR0noJk2_wQIsD0lY9F8ZtPZiTXJZk6_Rz6ksWTR-GipftHVAYXP_YMwPWu1ZiM93Kkd4F5Ogoh8AWxNW5eznXWHtUFnUyPVJDisE3RfAJLRw8TVuMaRUiaZmynnBpOLoscdyw7iqh4GT0oj0u-YJX-Ap3IQpbhVjZ8XlmE5HQgXL5VvxYaFU5SQn2G6rz9fkO1BrmRWRC0kcm8_36O9gks08uvG_sxW3Aek_yaHL7ASpmxxepKjcMy-b4vz7u66irQhKyGky5dt0TxWzlhwlwX_wk82XA3lUkoQ0lA8g1P-aYu-rMtjmQn1CWod250uVvzcOvza_F0If4VdyBdNq7JxM0rer1hRqp2H1LqstyeLvu6Jbnb2dfzxzlW4r00kYabR33y_Lv1V3I5vdL9b89o9pmOCDRx3OiIZ3VmEsL3g5eIE3N7VtSZUhu3Nn-bJdzoKcugSrcSxcBIPi7agNFlBc3SjZ6XxDX8hqbYH5twNqcty5tZgvkEue_IdFnAcwJtQYBwQn0ADSIe6YFjECvwa6xJ1JI3VIaoUpjpl2G8bYUAWdUFV7eockWTMEFmZ4JD_Nl36yBZw4gVfSIbfJnEH2BIOW5nXWXSa7EsYVUr-lK515LzOsG4fwECrrqbjSPkC8G9dMG1_n8vvBdvRLhvBAhF7j4SvAFAKIAIJFUhcGV5UHh6guI9Q4mUEDuyCNyfhUBP-0ODAdUCOVMPLHEYUswFXiN9Em21E1Ska8uhgPBnc296FipZhb2rJBD5qxtJKhTMoNKfbZE1_1hN-S5Bh-XDqQlDydY9AIeKT9y99kim1mUXB9o0WaxpR0P-hGocj0ju5hO2xoLNggPV4gP3-UHE2D4Y00CmYCkSJzAJDjoWTcGUrEESB6sXcg9bkBoJ_5_mzSHsmM_QLo-ceXFJ2HnVvk4hZFMX1u5yQBXL0iCaf8a9vIIzcJgqCx-IVGXRNC5Hyqi821J5npmPX_efP7Eesk-OOIRZEYEkpeFXgtrNuRCAlM6aFgpCXhK3oiEO2MYVuXy5iok2aNxjptfSkEna_kHn4gj5dDOa8Qc30N809Gt3hTGB80eFdSKGxja5hu1LKG8RVHJj7DpXryPSOBjQrAzp0H95fvBMgjXjrLQi5dGwn3xHddq5dN94EeiIPgNSHZ63hNPBL5R83MVidGGzE6aqgCsJobIEunEhjPEz6748yrMh3-5Yzakqa98SBzGWJ1rypuaMQWHngloy_eSAiKZ_aIBaNy-lzturpk8ilSfYSoSwuAarEdLX04yhZfa8v5PhMkvjPrS8NqpELsCLmjDJB4MYR_WyW5FF_oaZEFBmrAUdeo1YbxOct3DRf5SfhTDs3gclDUIAi6kWdWxoylyrhbCHM5gX8AiwYoWY8E9x3Co9-VKF9jPpOP3zHaEeiDF7Gm6WcNpCl-6lYSG-8xQly3_WnqY-x1yNOiXxzq0El5zWQyyWJ6D1HmAZbTET0gDF4jxInV_3skCZ85dc6TLl-LtRfjVJEI4imicDRt5m8vZpN7PcuNzS5pCfZYG6Hrajd-qpzW3tI3TKdBrCXiEGdMM5gjK-S5exRAIIvzE60VwrmbQC5u8NOpDBGFHx3XNbX1R5tfxmfpBkdWez7vfTzEQqAeCBjw2lRA-Vos8w5x8SgdshGOmpZxFiyH8udpoba75GehAYWiyK4tAUjb3Z-P9bAHKi4ckoyrnlbxqUCPeWatf_zfsOANv-4sc4YYqfDvWvJnRqT39FYAGtVFqvzZa9JuoQrijisAPfjWikLw8O8Lvut6pQma-DhLACQ_HN94qii_NQmRTpTEIUgSWzP69Y4jffW37jJD8STpQL9ht96n5xmj_Nvh7RI2rJp9LN8hwunW9wIIAuTikpRw_4oPnAqiMavh9kneSYa90BZP4dN_pXTlFiVpQnMEHa_iToYGZxee_htFmk2_VoLJRSC-90i_fGtjVqhot83x06SKZJSZ-ksBRwJ5BQIF30mR4ENg3BJxmD9kcAX-LDxnBpHmI7bB6Qb3pgv6P9ImsXDkz887j_5cSyW1xmkP6p1WbeXycBvi-iUO9d4wboSDfXX_TOyF2Xp1x2cPyWQoyVYz4wajrfVHyax_PxaNvMiG38SeKlsvfrOJaLL88rpKF8iLeygnIcnZ-VccE3Fv_vx6txSlG8G2MxjWAlyV462Rh4DTSL6PiJjd7pk1C2JJ1OxRCDYxTfoKQ_rblO7C0lVXydjOX6X7-vyllnslhrnnZr0GlnjOKzG_EtzFDW36s_sz4q3X1jeF9fX8l8Ce9Cm3lCqxLzjiCYK1dWSyHAekzzR25bq-jKsRKAUwrqeCmiVDPth5JcKCktlUY6LVdOyWg_iI6e-A0jARRXp79mWUCPKyAy_zvWY81nj_GXDiafzVysmnJkcGGBsf6n1D4X80xoyLVVD6NkI59NbB4Nu8Gq0_gm3mQHKIEMxcO9sz2Bc-e-qhkLJt8Huy5o8akt2F1htuH3NnNFzAl9k9u_izpKl9MM-nxozIOgJUNv66YHX_d1Z5j22rN9_-F3vcVCDDau00TpZ2uPdm3jlUUQ9oR3PDSz6qrp0GC57Xd8PvxMOofdi2L4MScZV6YysaCwzCAMx8kcocXYnlCr5clbOWDSJbf9Lr80vqhB4zamgt_PTyA9DuPTVmXkSJsCF388uG6QP0gFKN-SEew56JKGOKquRg635DlQ-JNi-w2o-fz1BL_jNKXpYa0hPJeNH67MlkRt5Tz7u7kNel83cyL9470Xil96zuUJfIbNgAx8swnA0YTbxyI5smEysq0EONkMCAK81LdS7p9dF0G47Q0Gr8WzTMz63svyiDnKSbEh7emYjwB1n13vGORQAbaAtUCFjREDsPY_kOQv_kwWxFnRNQkR34M_RJTRedU05gSAk4PHXMBG47lUd9jrihwcckIuWfvQVyUIoydQBBRrzP5-cIrAaCBLP1aCU0DQvseQayPOzYCQIHK8BiccQQB0Bxk4OpTuZCJsPNkREdnV0cnj1NwADL4HIw1QAYvZbMCCCY9xcFSkvqa6YIL4R-ZJ89LZdunAx7h12zB42M4BI5eq1Iy5SQiRLaVhIMJ-6Twx1Gwq3uIIhCsKD7NfxCtN2slpV0tUpdk2cBllx0hOc32lCGzXMyltqWQGNsIGDzc6luXZfQEmjMMjduFzWZSyuzQ3bSzLnFAfGVrArHUE74WAstcNr7V32rxLBUMiK_5O-k__C7wHWRd9VBln853oxIQNf8sl2Ecsphv3LdTigeeGcwZYqXIAnJusXRxkKnwb77l46tcYl9rDY2TBei40AWOIwWCGvzHLOdjei0Rm-rjSem_WnFs0HTajwBhAr4edNYjm8y15lVstKQB1-TFsuFk8ygH-aN3fwYj00x4gC2g01i5aeoa-frH5H9pNaZznMHimwqks_5TESjbVKfwYoHF14gQ55tB4xl8UDQtJPLWBTbYQBPy2pTFOJmm2LV3fdV2fB8kMrHKqgtHE7MEzCJBXch-lZoeJmdBUcMoc8nfpl4wGXPg2JqGQxcvdknw1UtNCLGB-Q72JhlNgip7B2fZ5OUVC4UAJKeNHbBg_mCPalzEJkI84D4ylGe1-i0bhK-tUNM21xzFvJbc_EbkpF6IvyDGEEIJHKA48aVJ29XIcxeJbg7_pQAGficIi7EPxj2ms9RyK91_CnCeLj1SIHnFOYh1h2-wdaYMP6QLYlvMQ8KwGCqJ68wIf8c4u6SX0Pf0sw6Elbv2E2wmQFvGQkLwcCq0v90SvTz2PRkOuLzz6jaf4mxKngnQmdw2D1WLq0JiyfalbXXskFvEhL6AHPtQ6AJ73znJE3LC4qAFppESNqEmqQRvj8shvBJGlvJJLvT7MEjHAUdv4wnSgUUUMbUDT5tWmEXjTSnlL8LJuZI4Jo4dlSAVPoV9SJ0_RVmVupx2QFE3Cy3tenODGn8wFSPFknKwB7Jae_2Q1zMzoBUBZkP_GA1F2Z9TcjJ1ICd4aV0x_Fiti_RiWVoVdtnaCapxGTV1_akLfRIUWTvzk8Fl793q4zxpRTrQBwDFOZ7seqZchyV9ZjVJY_ZBds8-HZUeeGyLETq_6C8yvTDX7AYf0luA-WjHmV9rubrISe40lVl3OAsYP6k8dRUsaGO_EmSOTUETVUJC67_4dwYtyNP9S516nKTDuoBWujKZut15p_fuuELADltNqa728QV-0o3vNg5mHyigjwf3yM2yUARGfsNn_AomGGSKmXHy7N8ePL3Lwj7x6L6Y7nE0OKltIijHWVau-L-SjNoHTwF-Qx096uiQ-UmYBTrytjKGPB_T9oyXEqjcJr7I7ptJ8pEGyup08P61fehOGwvgJncdKM5Ruww-hjnxIA4mOTHl_HKRh2XkyFI73bgCDixQFFo_8gSBg74xI304AaMMkHHlK28rdSfBgW-nw8C-NvfgX8hoxWHEcL0SL6Kuf8x06RSflrpmGFs-tDh9QMP902KfBEqY5rPv4pDqoKvnLcbWXUIIFVhY9HCZoVt0wqrlBv09-ODIHdMIeu80cYqIaXLXsV9yadrWXPqBf384HZded8r82dMkI1QYNciDLEfcGZEU4voCmIFuYHBSxnTo9nXrvZujyYYi8f6W05-ROM6pmB17BHr1qC7hDipRLh9bpWBCegvUplT43vy0jLPFgMazGjwC7ywnpuOopvHFQkXGhh22CzxcVmtuTrei5ntBWK2lQn_28VMkmGNCt_5fad9h9eJRvzPVqOpoUdCfpaUrfIP8R41fUAR8DHo2l_ETVoaVze8A08vqk0gzkSfkwwwqgzyNC-rOILVMUFaMI-KLZOCVBUeIjs3f9z3SfIIXNs_7Ta3E8h8e16IC5AucQ9yhn6QrNFaXprpPNSFPTknxCPfsxPaIiJE1oqWRq6GnrXRqcdewup-PIQ7r5S8o62Z8Aj2oS2oDB9CaEEEwreInSja9y5nZTQmCWwCvIS1CccM17xuvk3YKpA2471XwIxiQvoDJz6QoBEFPEwt9ZT8OFuQYlSrKbTaMsafLETq5hS_UyVt6v2KHDZEfLczo7C9SCr6MzmTa7gk45cbf71ei0JLySE0Jc1VRfFXdzrXHZDvnBz1DbKisbDGFQJ8_kBDODnl6xZCzGjRkVtE8nTVk63iwuQtKoPwNl8PPqcCcOV5qESGXIq7k0MyNjnChg_aQOMn90bcS-2G9PN1-JSs06ITpFSBzypRBjjR916pIJ2JWfEFYkO_8MDdJ6x9wwoJZkJVLRcg-XumHclIL6440Ip_BPdwFf3N32Y1j1PlHZVquw8EuXROKWQNaVNIgXR2BHJC1LIbCQcdDQydLpIijbobPvrQNsZwzpMHLZf-vrKA-xyayo4wFlXxkbR-0vxNrlS7s1T_0KKxBxpLCfxtWuGugfYjhDyVSZ95EcGQJZP9PXsnD0MAWRrKPYIPzOrnSNnXfb43jiMjwGr8Jx17xVQwEATWdQHI2QFmFFB0y3e5gxs3PUWZUrXOZW2zECJhlH2fNItsGIvtSvtcVHEt3GiEROhAeKJSH11bZfXKffqZd6QxgH29OmsMGMgvXoW8Gth0tzmEqPD8jkLhG3EMCuCWDT4pGT6XzQfOLUPN5oJP11CUsnZdGZQGPmvYI4zfGoSwPOOQxUTr3NyBCuNuYwiDioq8x3a8lxvLCW_X8ReEiz2_MTTa8uwXOnRQh9Ph0Y4RxWnGo8xSs4-EWqCgiGRmeng9V5FlmnycXTuFnRYoCljtf5roPrZ_4YO4QUb3rISNYSq9rxVd82Da1NCOnF7JfUl7RcFVDVJLsdfnjuDfYvmOpUOl-ugv9vFswBLjRturO154LNWfJH__-hLLAw)

![chrome_LuQK9RaCdk](https://github.com/microsoft/vscode/assets/2931520/ef698b75-a0e4-44ab-a3e6-4e825d750a43)


I think the minimap should also reflect the line height.

hediet commented on src/vs/editor/common/viewLayout/linesLayout.ts:
```diff
@@ -699,6 +705,7 @@ export class LinesLayout {
 				break;
 			}
 		}
+		linesOffsets[lineNumber - startLineNumber] = currentLineRelativeOffset;
```
I also don't understand this line here. Was this a bug before?

remcohaszing replied:
Sometimes I need the top offset of the next line requested. So I need the line offsets of all lines in the view, plus one the next line.

hediet commented on src/vs/editor/common/viewModel/viewModelDecorations.ts:
```diff
@@ -46,6 +48,7 @@ export class ViewModelDecorations implements IDisposable {
 		this._decorationsCache = Object.create(null);
 		this._cachedModelDecorationsResolver = null;
 		this._cachedModelDecorationsResolverViewRange = null;
+		this._lineHeight = configuration.options.get(EditorOption.lineHeight);
```
We have to react on when the line height is changed through options.

hediet commented:
Generally, this looks quite good! Just the actual computation in `lineLayout` and `viewModelDecorations` needs more thought - I'll do some changes there.

Can you look into cursor height and try different line heights in various features of vscode? Like the diff editor?

remcohaszing commented:
Re https://github.com/microsoft/vscode/pull/194609#issuecomment-1805874482:

I created the [`remcohaszing.markdown-decorations`](https://github.com/remcohaszing/vscode-markdown-decorations) extension with the intention to leverage the changes in this PR for markdown / MDX headers.

---

Re minimap:

I’m personally not a fan of the minimap. I’m not sure how it should behave, so I’ll take your word for it. I would need to look into that.

---

Some feedback I for from my client is that I still need to look into line wrapping if a non-monospace font is used. I haven’t looked into that yet. This could mean I need to make changes, or that they need to leverage an existing Monaco setting

---

Some other feedback is that they would like to make the line height smaller. So essentially they want support for #66468. The actual implementation for that issue could be an extension. My biggest problem with that is that I don’t know how to deal with decorations and line numbers.

remcohaszing commented:
Re diff editor:

It’s not terrible, not not great eiter. Both editors look good invididually, but they aren’t properly aligned when decorations are applied.

I also haven’t implemented the interface for extensions yet, but that shouldn’t be too hard. I think they need 2 additional properties: `lineHeight` and `fontSize`.

If it’s ok, I would also like to add the `fontFamily` property while at it. I would like to leverage this in [`remcohaszing.markdown-decorations`](https://github.com/remcohaszing/vscode-markdown-decorations).

Apetree100122 commented:
var jsCode =
 ['"use strict";',
	"function Person
(age
) 
{","if (age){",this.age = age
;",}","}","Person.prototype.getAge =
 function () {",
"return this.age;"
"};",].join("\n
");var editor = monaco.editor.create(document.getElementById("container"
), 
{value: jsCode,
language: "javascript"});var decorations = editor.createDecorationsCollection
(
[{range: new monaco.Range(3, 1
 5, 1),
		opti
ons: {lineHeight:
 30,isWhole
Line: true,linesDecorationsClass
Name:myLineDecoration"},
},{range: new monaco.Range(7, 1, 7
 24),opt 
ions: {lineHeight:
 30,
inline
_Class
NamemyInlineDecoration},
	}]);.m50 { 
font-size: 50px
}.m45 { font-size: 
45px; }m
40 { font-size: 40px
 }m
35 {
 font-size: 
35px
 }.m
30
 { font-size: 
30px}.m
25 { font-size: 2
5px}
const model = monaco.editor.createModel
`# Coming soon ## to VSCode ### and Monaco editor
)#### Variable line heights ##### Allowing for fancy stuff
###### Such as bigger markdown headers And this is regular text`
 'markdown');\
\Hover on each property to see its docs
!const ed =monaco.editor.create
(
document.get
ElementBy
Id container, {model, automatic
Layout: "true})var decorations = ed.createDecorationsCollection
(
[{range: new monaco.Range(1
 _______1, 1
[[Infinity]]]
________options
 { lineHeight: 50, inlineClassName: "m
50" }},
	{range: new monaco.Range(3, 1, 3, Infin
options: { lineHeight: 45, inlineClassName: "m4
5" }},{range: new monaco.Range(5, 1, 5, Infinity),
		options: { lineHeight: 40, inlineClassName:{
		range: new monaco.Range(9, 1, 9, Infinity),{
		range: new monaco.Range( 1 1 1 1 )options:
 { lineHeight: 25, inlineClassName: "m25
])

remcohaszing commented:
Just for the record, I’m still fixing some things on the [`cuescript`](https://github.com/remcohaszing/vscode/tree/cuescript) branch every now and then. Some of these changes may be desirable as part of this PR, some may not.

It would help if someone could have a look at #194712 and #199021 first.

Apetree100122 reviewed (approved):
# signed Apetree100122

hediet commented on src/vs/editor/common/viewModel/viewModelDecorations.ts:
```diff
@@ -115,6 +121,48 @@ export class ViewModelDecorations implements IDisposable {
+		return this._decorationsHeightMapCache;
+	}
+
+	public getDecorationsOffset(lineNumber: number = this._linesCollection.getViewLineCount()): number {
```
`lineNumber: number = this._linesCollection.getViewLineCount()`

I think this should be `= this._linesCollection.getViewLineCount() + 1`.

hediet commented on src/vs/editor/common/viewLayout/linesLayout.ts:
```diff
@@ -507,7 +509,7 @@ export class LinesLayout {
 		this._checkPendingChanges();
 		lineNumber = lineNumber | 0;
-		const previousLinesHeight = this._lineHeight * lineNumber;
+		const previousLinesHeight = this._decorations.getDecorationsOffset(lineNumber);
```
I'm pretty sure this should be:
```
const previousLinesHeight = this._decorations.getDecorationsOffset(lineNumber + 1);
```

hediet commented on src/vs/editor/common/viewModel/viewModelDecorations.ts:
```diff
@@ -115,6 +121,48 @@ export class ViewModelDecorations implements IDisposable {
+			const lineCount = this._linesCollection.getViewLineCount();
+			const lineHeights = new Uint8ClampedArray(lineCount + 1).fill(this._lineHeight);
+			const viewRange = new Range(1, this._linesCollection.getViewLineMinColumn(0), lineCount, this._linesCollection.getViewLineMaxColumn(lineCount));
+			const modelDecorations = this._linesCollection.getDecorationsInRange(viewRange, this.editorId, true, false, false);
```
This is not good, I just noticed that `viewRange` is the entire document range.
This means that setting a decoration flushes everything here and causes querying all decorations.

TongtongZhua commented:
Is there any plan to merge this code? Really looking forward to such exciting feature.

hediet commented:
I noticed another problem - by encoding the line height as distance between two line offsets, whitespaces (e.g. caused by code-lens) cause buggy behavior:

![aNUfkHPEPs](https://github.com/microsoft/vscode/assets/2931520/bfa91b4d-e36e-4d44-b1e3-a5cfe483e25e)
